### PR TITLE
local models

### PR DESCRIPTION
**Pull request by FelipeAdachi**
Langkit downloads models from HF Hub, which will hit an error in network-restricted environments. Even though modules such as `toxicity`,`themes`, and `input_output` allow passing the path of local model, the auto-initialization on import will hit an error before we're able to pass the local path in the config.

To deal with this, this change lazy initializes the models so it will fetch them when it's actually needed, when the udf is called. When using local model, one can do as below:

```python
from langkit import LangKitConfig

local_config = LangKitConfig(toxicity_model_path="local-toxicity-model",
              transformer_name="local-sentence-transformers")

toxicity.init(config=local_config)
themes.init(config=local_config)
input_output.init(config=local_config)
```

provided the supported models were already downloaded and stored in `local-toxicity-model` and `local-sentence-transformers`. For reference, these are the HF models that are currently being used in Langkit:

`martin-ha/toxic-comment-model`

`sentence-transformers/all-MiniLM-L6-v2`

### CODE REVIEW

File: langkit/examples/Local_Models.ipynb [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Using Langkit with Local Models\n",
+    "\n",
+    "[![Open in Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/whylabs/LanguageToolkit/blob/main/langkit/examples/Local_Models.ipynb)\n",
+    "\n",
+    "Some of the Langkit modules downloads models from the internet. This is not always possible, for example, when running in an environment without internet access. In this example, we will show how you can use Langkit with models stored locally."
```
Comment (richard-rogers) [inline langkit/examples/Local_Models.ipynb:19]:
```suggestion
    "Some of the Langkit modules download models from the internet. This is not always possible, for example, when running in an environment without internet access. In this example, we will show how you can use Langkit with models stored locally."
```

File: langkit/examples/Local_Models.ipynb [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+    "toxicity.init(config=local_config)\n",
+    "themes.init(config=local_config)\n",
+    "input_output.init(config=local_config)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "If, for example, we want a local version for the `llm_metrics` module, we also need to import `textstat`, `regexes`, and `sentiment`. `regexes` and `textstat` are lightweight models and doesn't require external artifacts, so we can use them in a network restricted environment. `sentiment`, however, downloads artifacts from the internet, so let's replace it with `vader_sentiment`, which will yield the same results as `sentiment`, with the benefit of not requiring downloading artifacts at runtime."
```
Comment (richard-rogers) [inline langkit/examples/Local_Models.ipynb:108]:
```suggestion
    "If, for example, we want a local version for the `llm_metrics` module, we also need to import `textstat`, `regexes`, and `sentiment`. `regexes` and `textstat` are lightweight models and don't require external artifacts, so we can use them in a network restricted environment. `sentiment`, however, downloads artifacts from the internet, so let's replace it with `vader_sentiment`, which will yield the same results as `sentiment`, with the benefit of not requiring downloading artifacts at runtime."
```

File: langkit/utils.py [hunk 3]
```diff
@@ -1,11 +1,38 @@
...
+        self.__init = init
+
+    def value(self, arg: In) -> Out:
+        if arg not in self.__cache:
+            self.__cache[arg] = self.__init(arg)
+
+        return self.__cache[arg]
+
+
+class LazyInit(Generic[Out]):
```
Comment (jamie256) [inline langkit/utils.py:25]: Can we prefix with _ to signify these are internal.

File: langkit/utils.py [hunk 4]
```diff
@@ -1,11 +1,38 @@
...
-from typing import Optional
+from typing import Optional, Generic, Callable, Dict, TypeVar
 from langkit import lang_config
 import string
 import random
 import functools
 import warnings
 
+Out = TypeVar("Out")
+In = TypeVar("In")
```
Comment (jamie256) [inline langkit/utils.py:10]: functools cache decorator?

File: langkit/toxicity.py [hunk 5]
```diff
@@ -14,47 +19,60 @@
...
         result[0]["score"] if result[0]["label"] == "toxic" else 1 - result[0]["score"]
     )
 
 
+def init_model():
+    if _model_path is None:
+        raise ValueError("Must initialize model path before calling toxicity!")
+    _model.value
+    _toxicity_tokenizer.value
+    _toxicity_pipeline.value
```
Comment (jamie256) [inline langkit/toxicity.py:57]: Curios if this is noticeably slower or not?

File: langkit/transformer.py [hunk 6]
```diff
@@ -28,7 +33,6 @@ def __init__(
         self,
         transformer_name: Optional[str],
         custom_encoder: Optional[Callable[[List[str]], Any]],
-        veto_cuda: bool = False,
```
Comment (jamie256) [inline langkit/transformer.py:31]: Do we need to remove this? What's the context?

**Review by jamie256 — COMMENTED** [review]
Left some preliminary comments.

File: langkit/toxicity.py (hunk 5 above)
Comment (FelipeAdachi) [reply langkit/toxicity.py:57]: I did a very simple test, iterating on 5k samples (single row extraction), and `init_model()` takes, on average 0.0048 ms (4.8e-6 seconds).

File: langkit/utils.py (hunk 3 above)
Comment (FelipeAdachi) [reply langkit/utils.py]: removed those, not relevant anymore

File: langkit/transformer.py (hunk 6 above)
Comment (FelipeAdachi) [reply langkit/transformer.py]: put this back in

File: langkit/utils.py (hunk 4 above)
Comment (FelipeAdachi) [reply langkit/utils.py]: added cache decorator

**Review by jamie256 — APPROVED** [review]
LGTM!

Let's file an issue for local model support on other toxicity models.

Also consider a post_init() or other way to trigger initialization outside of an actual request. We could document that you can call predict on these models as a first time call that should trigger any downloads or initialization.

**Pull request merged**